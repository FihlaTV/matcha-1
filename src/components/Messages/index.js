// @flow
import React from 'react';
import { Text, View, StyleSheet, Image } from 'react-native';
import type { MessageType } from 'src/types/Message';
import { MARGINS, COLORS } from 'src/constants/design';
import moment from 'moment';
import { injectIntl, type IntlShape } from 'react-intl';
import LoadingHOC from 'src/components/HOC/LoaderHOC';

const styles = StyleSheet.create({
  leftBubble: {
    alignSelf: 'flex-start',
    flexDirection: 'row',
    marginTop: MARGINS.TINY,
  },
  leftBubbleText: {
    alignSelf: 'flex-start',
    backgroundColor: COLORS.GREY_LIGHTEST,
    borderRadius: 20,
    flexDirection: 'row',
    flexWrap: 'wrap',
    paddingHorizontal: MARGINS.SMALL,
    paddingVertical: MARGINS.TINY,
  },
  messageContainer: {
    alignItems: 'center',
    backgroundColor: COLORS.BACKGROUND,
    padding: MARGINS.SMALL,
  },
  messageProfile: {
    backgroundColor: COLORS.GREY_DARK,
    borderRadius: 20,
    height: 40,
    marginRight: MARGINS.SMALL,
    width: 40,
  },
  messagesContainer: {
    flex: 1,
  },
  photoProfile: {
    backgroundColor: COLORS.GREY_DARK,
    borderRadius: 10,
    height: 20,
    marginRight: MARGINS.SMALL,
    marginTop: MARGINS.TINY,
    width: 20,
  },
  rightBubble: {
    alignSelf: 'flex-end',
    backgroundColor: COLORS.BLUE_LIGHT,
    borderRadius: 20,
    flexDirection: 'row',
    marginTop: MARGINS.TINY,
    paddingHorizontal: MARGINS.SMALL,
    paddingVertical: MARGINS.TINY,
  },
  timeText: {
    color: COLORS.GREY_DARK,
    fontSize: 12,
  },
});

type MessageProps = {
  message: MessageType,
};

const Message = (props: MessageProps) => (
  <View style={styles.messageContainer}>
    <Text style={styles.timeText}>
      {
        moment(Date.now()).diff(moment(props.message.createdAt), 'days') > 5 // eslint-disable-line
          ? moment(props.message.createdAt).format('dddd, MMM Do YYYY, h:mm:ss')
          : moment(props.message.createdAt).fromNow()
      }
    </Text>
    {
      props.message.from.username === 'admin' ?
        <View style={styles.rightBubble}>
          <Text style={{ color: COLORS.WHITE }}>{props.message.message}</Text>
        </View> :
        <View style={styles.leftBubble}>
          {/* eslint-disable-next-line */}
          <Image style={styles.photoProfile} source={{ uri: props.message.from.picture }} defaultSource={require('src/assets/images/no-pic.jpg')} />
          <View style={styles.leftBubbleText}>
            <Text style={{ flexWrap: 'wrap', maxWidth: 250 }}>{props.message.message}</Text>
          </View>
        </View>
    }
  </View>
);

type MessagesProps = {
  intl: IntlShape,
  messages: Array<MessageType>,
};

const Messages = (props: MessagesProps) => (
  <View style={styles.messagesContainer}>
    {props.messages.length > 0 &&
      <View style={styles.messageContainer}>
        <Text style={styles.timeText}>
          {moment(props.messages[0].createdAt).format('dddd, MMM Do YYYY, h:mm:ss')}
        </Text>
        <Text style={{ color: COLORS.GREY_DARK, marginTop: MARGINS.TINY, textAlign: 'center' }}>
          {props.intl.formatMessage({ id: 'Messages.match' })}
        </Text>
      </View>}
    {props.messages.map((message) =>
      (
        !message.match.data[0]
          ? <Message key={message.id} message={message} />
          : null
      ))
    }
  </View>
);

export default injectIntl(LoadingHOC('messages')(Messages));
export { Messages, Message };
