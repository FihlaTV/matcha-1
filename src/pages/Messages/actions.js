import type { ThreadType } from 'app/types/Thread';
import type { MessageType } from 'app/types/Message';

export const LOAD_THREADS = '@MESSAGES_REDUCER/LOAD_THREADS';
export const ON_LOAD_THREADS = '@MESSAGES_SAGA/ON_LOAD_THREADS';
export const SELECT_THREAD = '@MESSAGES_REDUCER/SELECT_THREAD';
export const ON_SELECT_THREAD = '@MESSAGES_REDUCER/ON_SELECT_THREAD';
export const LOAD_MESSAGES = '@MESSAGES_REDUCER/LOAD_MESSAGES';
export const ON_LOAD_MESSAGES = '@MESSAGES_SAGA/ON_LOAD_MESSAGES';

export const loadThreads = (threads: Array<ThreadType>) => ({
  threads,
  type: LOAD_THREADS,
});

export const onLoadThreads = () => ({
  type: ON_LOAD_THREADS,
});

export const selectThread = (thread: string) => ({
  thread,
  type: SELECT_THREAD,
});

export const onSelectThread = (thread: string) => ({
  thread,
  type: ON_SELECT_THREAD,
});

export const loadMessages = (messages: Array<MessageType>) => ({
  messages,
  type: LOAD_MESSAGES,
});

export const onLoadMessages = (thread: string) => ({
  thread,
  type: ON_LOAD_MESSAGES,
});
