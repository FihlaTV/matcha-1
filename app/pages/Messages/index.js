import React, { Component } from 'react';
import { connect } from 'react-redux';
import { ScrollView, StyleSheet } from 'react-native';
import Layout from 'app/components/Layout/Basic';
import Threads from 'app/components/Threads';
import { onLoadThreads, onSelectThread } from 'app/pages/Messages/actions';
import { selectThreads } from 'app/pages/Messages/selector';
import type { ThreadType } from 'app/types/Thread';
import { COLORS, MARGINS } from 'app/constants/design';
import { Actions } from 'react-native-router-flux';

const styles = StyleSheet.create({
  mainPanel: {
    backgroundColor: COLORS.BACKGROUND,
    flex: 1,
    margin: MARGINS.TINY,
  },
});

type Props = {
  threads: Array<ThreadType>,
  onSelectThread: (string) => {},
};

class Messages extends Component<Props> {
  componentWillMount() {
    this.props.onLoadThreads();
  }

  render() {
    return (
      <Layout>
        <ScrollView style={styles.mainPanel}>
          <Threads
            threads={this.props.threads}
            onSelectThread={(thread) => { this.props.onSelectThread(thread); Actions.push('conversation'); }}
          />
        </ScrollView>
      </Layout>
    );
  }
}

const mapStateToProps = (state) => ({
  threads: selectThreads(state),
});

const mapDispatchToProps = {
  onLoadThreads,
  onSelectThread,
};

export default connect(mapStateToProps, mapDispatchToProps)(Messages);
