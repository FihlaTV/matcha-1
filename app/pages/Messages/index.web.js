import React, { Component } from 'react';
import { connect } from 'react-redux';
import { ScrollView, StyleSheet } from 'react-native';
import Layout from 'app/components/Layout/Basic';
import Threads from 'app/components/Threads';
import MessagesComponent from 'app/components/Messages';
import { onLoadThreads, onSelectThread } from 'app/pages/Messages/actions';
import { selectThreads, selectMessages, selectThread } from 'app/pages/Messages/selector';
import type { ThreadType } from 'app/types/Thread';
import type { MessageType } from 'app/types/Message';
import { COLORS, MARGINS } from 'app/constants/design';
import ProfileComponent from 'app/components/Users/Profile';
import type { PhotoType } from 'app/types/Photo';
import { selectPhotosUser } from 'app/pages/Feed/selector';

const styles = StyleSheet.create({
  leftPanel: {
    backgroundColor: COLORS.BACKGROUND,
    flex: 0,
    margin: MARGINS.TINY,
    width: 325,
  },
  mainPanel: {
    backgroundColor: COLORS.BACKGROUND,
    flex: 1,
    marginVertical: MARGINS.TINY,
    minWidth: 325,
  },
  rightPanel: {
    backgroundColor: COLORS.BACKGROUND,
    flex: 0,
    margin: MARGINS.TINY,
  },
});

type Props = {
  threads: Array<ThreadType>,
  messages: Array<MessageType>,
  onSelectThread: (string) => {},
  selectedThread: ThreadType,
  photosUser: Array<PhotoType>,
};

class Messages extends Component<Props> {
  componentWillMount() {
    this.props.onLoadThreads();
  }

  render() {
    return (
      <Layout style={{ flexDirection: 'row' }}>
        <ScrollView style={styles.leftPanel}>
          <Threads threads={this.props.threads} onSelectThread={this.props.onSelectThread} />
        </ScrollView>
        <ScrollView style={styles.mainPanel}>
          <MessagesComponent messages={this.props.messages} />
        </ScrollView>
        { this.props.selectedThread &&
          <ProfileComponent user={this.props.selectedThread.user2} photos={this.props.photosUser} />
        }
      </Layout>
    );
  }
}

const mapStateToProps = (state) => ({
  messages: selectMessages(state),
  photosUser: selectPhotosUser(state),
  selectedThread: selectThread(state),
  threads: selectThreads(state),
});

const mapDispatchToProps = {
  onLoadThreads,
  onSelectThread,
};

export default connect(mapStateToProps, mapDispatchToProps)(Messages);
