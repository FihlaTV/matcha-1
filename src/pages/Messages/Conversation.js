import React from 'react';
import { connect } from 'react-redux';
import { ScrollView, StyleSheet } from 'react-native';
import Layout from 'app/components/Layout/Basic';
import Messages from 'app/components/Messages';
import { selectMessages, selectThread } from 'app/pages/Messages/selector';
import type { MessageType } from 'app/types/Message';
import { COLORS, MARGINS } from 'app/constants/design';
import type { ThreadType } from 'app/types/Thread';
import Icon from 'react-native-vector-icons/FontAwesome';

const styles = StyleSheet.create({
  mainPanel: {
    backgroundColor: COLORS.BACKGROUND,
    flex: 1,
    margin: MARGINS.TINY,
  },
});

type Props = {
  messages: Array<MessageType>,
  selectedThread: ThreadType,
};

const Conversation = (props: Props) => (
  <Layout
    noTabs
    hasBack
    title={props.selectedThread ? props.selectedThread.user2.name : undefined}
    headerActions={[
      <Icon name="ellipsis-h" key="ellipsis-h" size={20} color={COLORS.BLUE_DARKER} />,
    ]}
  >
    <ScrollView style={styles.mainPanel}>
      <Messages messages={props.messages} />
    </ScrollView>
  </Layout>
);

const mapStateToProps = (state) => ({
  messages: selectMessages(state),
  selectedThread: selectThread(state),
});

export default connect(mapStateToProps)(Conversation);
export { Conversation };
