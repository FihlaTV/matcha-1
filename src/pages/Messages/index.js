import React, { Component } from 'react';
import { connect } from 'react-redux';
import { ScrollView } from 'react-native';
import Layout from 'app/components/Layout/Basic';
import Threads from 'app/components/Threads';
import { onLoadThreads } from 'app/pages/Messages/actions';
import { selectThreads } from 'app/pages/Messages/selector';
import type { Thread } from 'app/types/Thread';

type Props = {
  threads: Array<Thread>,
};

class Messages extends Component<Props> {
  componentWillMount() {
    this.props.onLoadThreads();
  }

  render() {
    return (
      <Layout>
        <ScrollView style={{ width: 300 }}>
          <Threads threads={this.props.threads} />
        </ScrollView>
      </Layout>
    );
  }
}

const mapStateToProps = (state) => ({
  threads: selectThreads(state),
});

const mapDispatchToProps = {
  onLoadThreads,
};

export default connect(mapStateToProps, mapDispatchToProps)(Messages);
export { Messages };
