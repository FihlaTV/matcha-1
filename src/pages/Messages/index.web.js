import React, { Component } from 'react';
import { connect } from 'react-redux';
import { ScrollView, StyleSheet } from 'react-native';
import Layout from 'app/components/Layout/Basic';
import Threads from 'app/components/Threads';
import MessagesComponent from 'app/components/Messages';
import { onLoadThreads, onSelectThread } from 'app/pages/Messages/actions';
import { selectThreads, selectMessages, selectThread } from 'app/pages/Messages/selector';
import type { Thread } from 'app/types/Thread';
import type { Message } from 'app/types/Message';
import { COLORS, MARGINS } from 'app/constants/design';
import { User } from 'app/components/Users';

const styles = StyleSheet.create({
  leftPanel: {
    backgroundColor: COLORS.BACKGROUND,
    flex: 0,
    margin: MARGINS.TINY,
    width: 325,
  },
  mainPanel: {
    backgroundColor: COLORS.BACKGROUND,
    flex: 1,
    marginVertical: MARGINS.TINY,
    minWidth: 325,
  },
  rightPanel: {
    backgroundColor: COLORS.BACKGROUND,
    flex: 0,
    margin: MARGINS.TINY,
  },
});

type Props = {
  threads: Array<Thread>,
  messages: Array<Message>,
  onSelectThread: (string) => {},
  selectedThread: Thread,
};

class Messages extends Component<Props> {
  componentWillMount() {
    this.props.onLoadThreads();
  }

  render() {
    return (
      <Layout style={{ flexDirection: 'row' }}>
        <ScrollView style={styles.leftPanel}>
          <Threads threads={this.props.threads} onSelectThread={this.props.onSelectThread} />
        </ScrollView>
        <ScrollView style={styles.mainPanel}>
          <MessagesComponent messages={this.props.messages} />
        </ScrollView>
        { this.props.selectedThread &&
          <ScrollView style={styles.rightPanel}>
            <User user={this.props.selectedThread.user2} />
          </ScrollView>
        }
      </Layout>
    );
  }
}

const mapStateToProps = (state) => ({
  messages: selectMessages(state),
  selectedThread: selectThread(state),
  threads: selectThreads(state),
});

const mapDispatchToProps = {
  onLoadThreads,
  onSelectThread,
};

export default connect(mapStateToProps, mapDispatchToProps)(Messages);
export { Messages };
