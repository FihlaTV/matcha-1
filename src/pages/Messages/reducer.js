import { LOAD_THREADS } from 'app/pages/Messages/actions';
import type { Thread } from 'app/types/Thread';

type State = {
  threads: Array<Thread>,
};

const initialState: State = {
  threads: [],
};

export default (state: State = initialState, action: any): State => {
  if (action.type === LOAD_THREADS) {
    return { ...state, threads: action.threads };
  }
  return state;
};
