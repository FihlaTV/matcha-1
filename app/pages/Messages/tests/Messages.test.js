import React from 'react';
import Messages from 'app/pages/Messages';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';

describe('Messages', () => {
  describe('render test', () => {
    it('should render', () => {
      const onLoadThreads = jest.fn();
      const wrapper = shallow(<Messages onLoadThreads={onLoadThreads} />).dive();
      expect(toJson(wrapper)).toMatchSnapshot();
      expect(onLoadThreads).toHaveBeenCalledTimes(1);
    });
  });
});
