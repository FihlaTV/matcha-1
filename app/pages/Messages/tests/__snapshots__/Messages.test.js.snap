// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Messages render test should render 1`] = `
<Layout>
  <ScrollView
    style={
      Object {
        "backgroundColor": "white",
        "flex": 1,
        "margin": 5,
      }
    }
  >
    <LoadingHOC(Threads)
      onSelectThread={[Function]}
    />
  </ScrollView>
</Layout>
`;
